File: python/linked_list/linked_list.py
class Element(object):
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    def append(self, new_element):
        current = self.head
        if self.head:
            while current.next:
                current = current.next
            current.next = new_element
        else:
            self.head = new_element

    def get_position(self, position):
        current = self.head
        for i in range(position - 1):
            current = current.next
        return current

    def insert(self, new_element, position):
        if position == 1:
            new_element.next = self.head
            self.head = new_element
        else:
            item = self.get_position(position - 1)
            new_element.next = item.next
            item.next = new_element
        pass


    def delete(self, value):
        if self.head.value == value:
            self.head = self.head.next
            return
        current = self.head
        while current.next:
            if current.next.value == value:
                current.next = current.next.next
            else:
                current = current.next
        pass

File: ruby/linked_list/linked_list.rb
class Element
  attr_accessor :value, :next

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedList
  attr_reader :head
  def initialize(head = nil)
    @head = head
  end

  def append(new_element)
    current = head
    if head
      while current.next do
        current = current.next
      end
      current.next = new_element
    else
      @head = new_element
    end
    self
  end

  def get_position(position)
    current = head
    for i in (1...position) do
      current = current.next
    end
    current
  end

  def insert(new_element, position)
    if position == 1
      new_element.next = head
      @head = new_element
    else
      previous = get_position(position - 1)
      new_element.next = previous.next
      previous.next = new_element
    end
  end

  def delete(value)
    if head.value == value
      @head = head.next
    else
      current = head
      while current.next do
        if current.next.value == value
          return current.next = current.next.next
        else
          current = current.next
        end
      end
    end
  end

  def insert_first(new_element)
    insert(new_element, 1)
  end

  def delete_first
    target = get_position(1)
    delete(target.value) if target
    target
  end
end
